vrouter.go

```diff
@@ -300,11 +309,85 @@ func (self *Vrouter) AddLocalEndpoint(endpoint OfnetEndpoint) error {
 		}
 	}
 
+	if endpoint.HostPvtIP.String() != "<nil>" && self.hostNATInfo.PortNo != 0 {
+		return self.setupHostNAT(endpoint, vrfmetadata, vrfmetadataMask)
+	}
+	return nil
+}
+
+func (self *Vrouter) setupHostNAT(endpoint OfnetEndpoint, vrfmetadata, vrfmetadataMask uint64) error {
+	// setup the host NAT flows
+	hsNAT, err := self.hostSNATTable.NewFlow(ofctrl.FlowMatch{
+		Priority:  FLOW_MATCH_PRIORITY,
+		Ethertype: 0x0800,
+		InputPort: endpoint.PortNo,
+	})
+
+	if err != nil {
+		log.Errorf("Error adding endpoint to host SNAT {%+v}. Err: %v", endpoint, err)
+		return err
+	}
+
+	hsNAT.SetIPField(endpoint.HostPvtIP, "Src")
+	hsNAT.SetMacDa(self.hostNATInfo.MacAddr)
+	natOut, err := self.ofSwitch.OutputPort(self.hostNATInfo.PortNo)
+	if err != nil {
+		return err
+	}
+	// Point the route at output port
+	err = hsNAT.Next(natOut)
+	if err != nil {
+		log.Errorf("Error installing host NAT out flow for endpoint: %+v. Err: %v", endpoint, err)
+		return err
+	}
+	snatFlowKey := endpoint.HostPvtIP.String() + ".snat"
+	self.hostNATFlowDB.Set(snatFlowKey, hsNAT)
+
+	hdNAT, err := self.hostDNATTable.NewFlow(ofctrl.FlowMatch{
+		Priority:  FLOW_MATCH_PRIORITY,
+		InputPort: self.hostNATInfo.PortNo,
+		Ethertype: 0x0800,
+		IpDa:      &endpoint.HostPvtIP,
+	})
+
+	if err != nil {
+		log.Errorf("Error adding endpoint to host DNAT {%+v}. Err: %v", endpoint, err)
+		return err
+	}
+
+	hdNAT.SetIPField(endpoint.IpAddr, "Dst")
+	eMac, err := net.ParseMAC(endpoint.MacAddrStr)
+	if err != nil {
+		log.Errorf("Error parsing ep MAC: %s", endpoint.MacAddrStr)
+		return err
+	}
+	hdNAT.SetMacDa(eMac)
+	hdNAT.SetMetadata(vrfmetadata, vrfmetadataMask)
+	// Point to SNAT table
+	err = hdNAT.Next(self.proxySNATTable)
+	if err != nil {
+		log.Errorf("Error installing host NAT in flow for endpoint: %+v. Err: %v", endpoint, err)
+		return err
+	}
+	dnatFlowKey := endpoint.HostPvtIP.String() + ".dnat"
+	self.hostNATFlowDB.Set(dnatFlowKey, hdNAT)
 	return nil
 }
 
 // Remove local endpoint
 func (self *Vrouter) RemoveLocalEndpoint(endpoint OfnetEndpoint) error {
+	// Remove host nat flows
+	dnatFlowKey := endpoint.HostPvtIP.String() + ".dnat"
+	hdFlow, _ := self.hostNATFlowDB.Get(dnatFlowKey)
+	if hdFlow != nil {
+		hdFlow.(*ofctrl.Flow).Delete()
+	}
+	snatFlowKey := endpoint.HostPvtIP.String() + ".snat"
+	hsFlow, _ := self.hostNATFlowDB.Get(snatFlowKey)
+	if hsFlow != nil {
+		hsFlow.(*ofctrl.Flow).Delete()
+	}
```

Comment:
@DivyaVavili Please see the three self.hostNATFlowDB.Remove at the end of the block.